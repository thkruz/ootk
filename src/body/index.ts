export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
