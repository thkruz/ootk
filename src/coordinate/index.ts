export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
