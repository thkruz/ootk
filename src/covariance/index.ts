export * from './CovarianceSample.js';
export * from './StateCovariance.js';
