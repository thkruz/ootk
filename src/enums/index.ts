export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
