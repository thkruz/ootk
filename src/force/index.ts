export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

