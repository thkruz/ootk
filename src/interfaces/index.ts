export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { TimeVariables } from './TimeVariables.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { ScenarioParams } from '../objects/Scenario.js';
export type { BaseObjectParams } from '../objects/base/BaseObject.js';
export type { GroundObjectParams } from '../objects/base/GroundObject.js';
export type { FacilityParams } from '../objects/base/Facility.js';
export type { SensorParams } from '../objects/attachable/SensorAdv.js';
export type { RectangularSensorParams } from '../objects/attachable/RectangularSensor.js';
