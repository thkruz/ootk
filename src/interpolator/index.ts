export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

