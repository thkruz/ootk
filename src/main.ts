/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
