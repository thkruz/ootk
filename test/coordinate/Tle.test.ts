import { ClassicalElements, EpochUTC, Kilometers, Radians, Tle, TleLine1, TleLine2 } from '../../src/main';

describe('Tle', () => {
  let tle: Tle;
  let exampleDate: Date;

  beforeEach(() => {
    // Sample TLE
    const tle1 = '1 56006U 23042W   24012.45049317  .00000296  00000-0  36967-4 0  9992' as TleLine1;
    const tle2 = '2 56006 143.0043  13.3620 0001137 267.5965  92.4747 15.02542972 44491' as TleLine2;
    const mockExampleDate = new Date(1705109326817);

    exampleDate = new Date(mockExampleDate.getTime());
    tle = new Tle(tle1, tle2);
  });

  // toString
  it('should return the TLE as a string', () => {
    expect(tle.toString()).toMatchSnapshot();
  });

  // semiMajorAxis
  it('should return the semi-major axis', () => {
    expect(tle.semimajorAxis).toMatchSnapshot();
  });

  // eccentricity
  it('should return the eccentricity', () => {
    expect(tle.eccentricity).toMatchSnapshot();
  });

  // inclination
  it('should return the inclination', () => {
    expect(tle.inclination).toMatchSnapshot();
  });

  // inclinationDeg
  it('should return the inclination in degrees', () => {
    expect(tle.inclinationDegrees).toMatchSnapshot();
  });

  // apogee
  it('should return the apogee', () => {
    expect(tle.apogee).toMatchSnapshot();
  });

  // perigee
  it('should return the perigee', () => {
    expect(tle.perigee).toMatchSnapshot();
  });

  // period
  it('should return the period', () => {
    expect(tle.period).toMatchSnapshot();
  });

  // propagate
  it('should propagate the TLE', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const propagated = tle.propagate(epoch);

    expect(propagated).toMatchSnapshot();
  });

  // state
  it('should return the state vector', () => {
    expect(tle.state).toMatchSnapshot();
  });

  // fromClassicalElements
  it('should create a TLE from classical orbital elements', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const elements = new ClassicalElements({
      epoch,
      semimajorAxis: 6943.547853722985 as Kilometers,
      eccentricity: 0.0011235968124658146,
      inclination: 0.7509087232045765 as Radians,
      rightAscension: 0.028239555738616327 as Radians,
      argPerigee: 2.5386411901807353 as Radians,
      trueAnomaly: 0.5931399364974058 as Radians,
    });
    const tle = Tle.fromClassicalElements(elements);

    expect(tle).toMatchSnapshot();
  });
});

// Generated by Qodo Gen

describe('convert6DigitToA5', () => {

  // Converts a 6-digit SCC number to a 5-digit Alpha-5 format by replacing first two digits with a letter
  it('should convert 6-digit SCC number to Alpha-5 format', () => {
    // Test cases with different first two digits
    expect(Tle.convert6DigitToA5('123456')).toBe('C3456');
    expect(Tle.convert6DigitToA5('183456')).toBe('J3456');
    expect(Tle.convert6DigitToA5('243456')).toBe('Q3456');
    expect(Tle.convert6DigitToA5('333456')).toBe('Z3456');
  });

  // Handles inputs with exactly 6 characters
  it('should return original input when not a 6-digit number or already Alpha-5', () => {
    // Less than 6 digits
    expect(Tle.convert6DigitToA5('12345')).toBe('12345');

    // Already Alpha-5 format
    expect(Tle.convert6DigitToA5('A2345')).toBe('A2345');
    expect(Tle.convert6DigitToA5('Z2345')).toBe('Z2345');

    // Edge case - exactly 6 characters but first is already a letter
    expect(Tle.convert6DigitToA5('B12345')).toBe('B12345');
  });
});

// Generated by Qodo Gen

describe('parseAll', () => {

  // Successfully parses both TLE lines and combines them into a single data object
  it('should combine data from both TLE lines into a single object', () => {
    const tleLine1 = '1 25544U 98067A   08264.51782528 -.00002182  00000-0 -11606-4 0  2927' as TleLine1;
    const tleLine2 = '2 25544 051.6416 247.4627 0006703 130.5360 325.0288 15.72125391563537' as TleLine2;

    const result = Tle.parseAll(tleLine1, tleLine2);

    expect(result.satNum).toBe(25544);
    expect(result.lineNumber1).toBe(1);
    expect(result.lineNumber2).toBe(2);
    expect(result.inclination).toBeCloseTo(51.6416);
    expect(result.rightAscension).toBeCloseTo(247.4627);
    expect(result.eccentricity).toBeCloseTo(0.0006703);
    expect(result.intlDes).toBe('1998-067A');
    expect(result.epochYear).toBe(8);
    expect(result.epochDay).toBeCloseTo(264.51782528);
    expect(result.meanMotion).toBeCloseTo(15.72125391);
  });

  // Throws error when satellite numbers don't match between lines
  it('should throw error when satellite numbers do not match between lines', () => {
    const tleLine1 = '1 25544U 98067A   08264.51782528 -.00002182  00000-0 -11606-4 0  2927' as TleLine1;
    const tleLine2 = '2 25545 051.6416 247.4627 0006703 130.5360 325.0288 15.72125391563537' as TleLine2;

    expect(() => {
      Tle.parseAll(tleLine1, tleLine2);
    }).toThrow('Satellite numbers do not match');
  });
});

// Generated by Qodo Gen

describe('parse', () => {

  // Successfully parses valid TLE lines and returns TleData object with orbital parameters
  it('should return a valid TleData object when given correct TLE lines', () => {
    const line1 = '1 25544U 98067A   20049.59954503  .00001714  00000-0  38792-4 0  9993' as TleLine1;
    const line2 = '2 25544  51.6454 207.5396 0005246  10.7845 349.3392 15.49172904214789' as TleLine2;

    const result = Tle.parse(line1, line2);

    expect(result).toBeDefined();
    expect(result.satNum).toBe(25544);
    expect(result.intlDes).toBe('1998-067A');
    expect(result.epochYear).toBe(20);
    expect(result.epochDay).toBe(49.59954503);
    expect(result.meanMoDev1).toBe(0.00001714);
    expect(result.bstar).toBeCloseTo(0.00003879, 8);
    expect(result.inclination).toBeCloseTo(51.6454, 4);
    expect(result.rightAscension).toBeCloseTo(207.5396, 4);
    expect(result.eccentricity).toBeCloseTo(0.0005246, 7);
    expect(result.argOfPerigee).toBeCloseTo(10.7845, 4);
    expect(result.meanAnomaly).toBeCloseTo(349.3392, 4);
    expect(result.meanMotion).toBeCloseTo(15.49172904, 8);
    expect(result.period).toBeCloseTo(92.95, 2);
  });

  // Throws error when satellite numbers don't match between line 1 and line 2
  it('should throw an error when satellite numbers don\'t match between lines', () => {
    const line1 = '1 25544U 98067A   20049.59954503  .00001714  00000-0  38792-4 0  9993' as TleLine1;
    const line2 = '2 25545  51.6454 207.5396 0005246  10.7845 349.3392 15.49172904214789' as TleLine2;

    expect(() => {
      Tle.parse(line1, line2);
    }).toThrow('Satellite numbers do not match');
  });
});
