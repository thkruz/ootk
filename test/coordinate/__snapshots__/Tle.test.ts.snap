// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tle should create a TLE from classical orbital elements 1`] = `
Tle {
  "epoch": EpochUTC {
    "posix": 1705109279.999904,
  },
  "line1": "1 00001U 58001A   24013.06111111 +.00000000 +00000+0 +00000-0 0 09990",
  "line2": "2 00001 043.0239 001.6180 0011236 145.4534 033.9125 15.00481387 00010",
  "satnum": 1,
  "satrec_": Object {
    "PInco": 0,
    "a": 1.088949028290718,
    "alta": 0.09017257141890522,
    "altp": 0.08772548516253043,
    "argpdot": 0.00007500841038277947,
    "argpo": 2.53864073822032,
    "atime": 0,
    "aycof": 0.0008000244899288449,
    "bstar": 0,
    "cc1": 0,
    "cc4": 4.3305287931362373e-7,
    "cc5": 0.00011839745575895689,
    "con41": 0.6033863216244524,
    "d2": 0,
    "d2201": 0,
    "d2211": 0,
    "d3": 0,
    "d3210": 0,
    "d3222": 0,
    "d4": 0,
    "d4410": 0,
    "d4422": 0,
    "d5220": 0,
    "d5232": 0,
    "d5421": 0,
    "d5433": 0,
    "dedt": 0,
    "del1": 0,
    "del2": 0,
    "del3": 0,
    "delmo": 1.0402336592605395,
    "didt": 0,
    "dmdt": 0,
    "dnodt": 0,
    "domdt": 0,
    "e3": 0,
    "ecco": 0.0011236,
    "ee2": 0,
    "epochdays": 13.06111111,
    "epochyr": 24,
    "error": 0,
    "eta": 0.01594821618248224,
    "gsto": 2.3394503579736323,
    "inclo": 0.7509087120487883,
    "init": false,
    "irez": 0,
    "isimp": false,
    "j2": 0.001082616,
    "j3": -0.00000253881,
    "j3oj2": -0.002345069720011528,
    "j4": -0.00000165597,
    "jdsatepoch": 2460322.56111111,
    "mdot": 0.06547085575553616,
    "method": "n",
    "mo": 0.591884782582577,
    "mus": 398600.8,
    "nddot": 0,
    "ndot": 0,
    "no": 0.06544379037496881,
    "nodecf": -0,
    "nodedot": -0.00006565190786601781,
    "nodeo": 0.028239427297268253,
    "omgcof": -0,
    "operationmode": "a",
    "peo": 0,
    "pgho": 0,
    "pho": 0,
    "plo": 0,
    "radiusearthkm": 6378.135,
    "satnum": "00001",
    "se2": 0,
    "se3": 0,
    "sgh2": 0,
    "sgh3": 0,
    "sgh4": 0,
    "sh2": 0,
    "sh3": 0,
    "si2": 0,
    "si3": 0,
    "sinmao": 0.5579261762949905,
    "sl2": 0,
    "sl3": 0,
    "sl4": 0,
    "t": 0,
    "t2cof": 0,
    "t3cof": 0,
    "t4cof": 0,
    "t5cof": 0,
    "tumin": 13.446839696959309,
    "vkmpersec": 7.905370510517634,
    "x1mth2": 0.4655378927918492,
    "x7thm1": 2.7412347504570556,
    "xfact": 0,
    "xgh2": 0,
    "xgh3": 0,
    "xgh4": 0,
    "xh2": 0,
    "xh3": 0,
    "xi2": 0,
    "xi3": 0,
    "xke": 0.07436691613317342,
    "xl2": 0,
    "xl3": 0,
    "xl4": 0,
    "xlamo": 0,
    "xlcof": 0.0015379049393216336,
    "xli": 0,
    "xmcof": -0,
    "xni": 0,
    "zmol": 0,
    "zmos": 0,
  },
}
`;

exports[`Tle should propagate the TLE 1`] = `
TEME {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": 3699.9683427262826,
    "y": -4340.514516979853,
    "z": 3939.5098576859164,
  },
  "velocity": Vector3D {
    "x": -6.276352177964951,
    "y": -3.990490995081803,
    "z": 1.4943962253306078,
  },
}
`;

exports[`Tle should return the TLE as a string 1`] = `
"1 56006U 23042W   24012.45049317  .00000296  00000-0  36967-4 0  9992
2 56006 143.0043  13.3620 0001137 267.5965  92.4747 15.02542972 44491"
`;

exports[`Tle should return the apogee 1`] = `6937.9838309710485`;

exports[`Tle should return the eccentricity 1`] = `0.0001137`;

exports[`Tle should return the inclination 1`] = `2.495895879509727`;

exports[`Tle should return the inclination in degrees 1`] = `143.0043`;

exports[`Tle should return the perigee 1`] = `6936.406312811701`;

exports[`Tle should return the period 1`] = `95.83752523784717`;

exports[`Tle should return the semi-major axis 1`] = `6937.195071891375`;

exports[`Tle should return the state vector 1`] = `
TEME {
  "epoch": EpochUTC {
    "posix": 1705056522.609888,
  },
  "position": Vector3D {
    "x": 6748.804592785081,
    "y": 1602.0774592717344,
    "z": 0.7217817565636419,
  },
  "velocity": Vector3D {
    "x": 1.3943615794564204,
    "y": -5.89340298337042,
    "z": 4.566020263208419,
  },
}
`;
